### ISSUE TITLE
Read timer goes negative when timer ends while device is sleeping

### ISSUE THREAD

**Issue by matejdro** [koreader/koreader#8031]
* KOReader version: v2021.07
* Device: Kobo Glo HD (kobo_alyssum)

#### Issue

![image](https://user-images.githubusercontent.com/507922/127187860-af22de54-58d4-4c55-a537-8f8fccbca97b.png)

#### Steps to reproduce

1. Set read timer for 30 minutes
2. Put device to sleep
3. Turn device back on after 2 hours (or any time greater than 30 minutes)
4. Open plugins menu
5. Note that read timer is negative

~If I leave it in the "negative state", sometimes device will display "Read timer alarm" notification about a minute after I bring it out of the standby, but other times it will just stay negative until I manually clear it.~

EDIT: If you wait 10-20 seconds, "Read timer alarm" notification will popup. But if you want to set the timer before that, you will be greeted with above picture.

Presumably, timer should clear itself if I turn the device and timer is overdue?

**Comment by NiLuJe** [comment]
Plug-in is mildly unmaintained ;o).

On the other hand, it's extremely simplistic IIRC, so it might be a low hanging branch...

**Comment by Frenzie** [comment]
It's always worked fine and always will, doesn't realistically need any maintaining.

Being negative sounds like a non-issue, but I suppose it should be caught to prevent a potential overflow issue many weeks or months later.

**Comment by matejdro** [comment]
It is just a mild annoyance, since I have to reset the timer every time it goes negative.

**Comment by Frenzie** [comment]
I can't reproduce your issue. It always goes off after waking up.

**Comment by Frenzie** [comment]
@NiLuJe Could the event loop get confused when something like https://github.com/koreader/koreader/issues/7092 is involved?

**Comment by matejdro** [comment]
Hm that's weird. Maybe it is something kobo-related?

Just for the record, clock appears fine for me (it always has accurate time, so it is not #7092)

**Comment by Frenzie** [comment]
Correction, I can reproduce. It must seems to be a "recent" regression.

**Comment by matejdro** [comment]
One thing that could be related here is that I've noticed that (in normal circumstances, without sleeping in between) read timer alarm is usually 30-60 seconds late on my device:

![image](https://user-images.githubusercontent.com/507922/127272295-6f6da7f3-6163-447f-8949-1debe7fd1f55.png)

Above timer was set to trigger at 22:00 sharp. Clock in the statusbar updated to 22:00 properly, but alarm was apparently 51 seconds late (it even states so in the alarm message).

**Comment by hius07** [comment]
It doesn't honour suspend, doesn't restart after exit, so needs some development.

<kbd>![timer](https://user-images.githubusercontent.com/62179190/127469903-e828a0b0-7997-4d70-91b7-8fa60b151969.jpg)</kbd>

**Comment by Frenzie** [comment]
> It doesn't honour suspend, doesn't restart after exit

The x minutes I set to remind myself to put down the book if I should lose track of time have no relevance whatsoever the next day. Very much a feature to me, not a bug.

And the same applies even if I set it for an hour or something and suspended it for 15 minutes somewhere in the middle. I wouldn't want it to randomly add another 15 minutes just because I suspended it.

**Comment by hius07** [comment]
A message over screensaver is not good.
The plugin should wait for resume and then show anything (adding or not suspend time might be an option in settings).

**Comment by Frenzie** [comment]
Emphasis on *over screensaver*. Hacking around suspend in that plugin and any other potential equivalent sounds like a lot of unnecessary complexity, especially when you consider it would have to be done individually to each and every comparable scenario.

**Comment by Frenzie** [comment]
Could be a regression from https://github.com/koreader/koreader/pull/7256?

**Comment by matejdro** [comment]
> The x minutes I set to remind myself to put down the book if I should lose track of time have no relevance whatsoever the next day. 

Agree completely. That is why even the way you reproduce (alarm appears in few seconds) is probably not ideal and alarm should just turn off itself after resuming too late.

**Comment by Frenzie** [comment]
I don't quite agree with that. That is, the program doesn't know it's useless to me by the point I see it. But if I set it for 15 minutes from now, that should be 15 minutes from now, not 5 minutes now, 15 hours sleep, and then suddenly *another* 10 minutes. It should be (more or less) instantaneously on wakeup.

If anything it should arguably schedule a wakeup to display the alarm (and from that perspective, yes, *over* the screensaver). Of course, that's very much not worth the effort.

In short I'm not at all convinced it's doing anything unexpected or undesirable, besides the thing in the OP obviously. ;-)

**Comment by matejdro** [comment]
I tested this issue out a bit and I think the main issue is that I'm too impatient.

It seems like every timer gets resolved, but the problem is that it might take 10-20 seconds for the popup to show up after sleep.

I'm usually setting timer immediately after starting to read, which usually results in me manually clearing the old timer (since 10-20 seconds have not passed yet) and setting the new one.

**Comment by hius07** [comment]
In any case these negative figures are confusing.

<kbd>![1](https://user-images.githubusercontent.com/62179190/127606194-565ec4f6-7d1c-4e59-b2db-3952060d2e12.png)</kbd>

<kbd>![2](https://user-images.githubusercontent.com/62179190/127606198-d5190b0c-2d5a-479d-81f1-0fe8aae91f03.png)</kbd>

**Comment by Frenzie** [comment]
@matejdro Oh wait, you said it does display the alarm and the -1100 minutes is just a minor *display* issue afterward. It still seems weird that it's delayed and I don't think it used to be.

**Comment by matejdro** [comment]
No, negative number is displayed BEFORE the alarm. After the alarm, negative disappears. But since I'm usually setting my alarms immediately after turning the device on, the delay in the alarm causes the issue.

**Comment by Frenzie** [comment]
Right, so the entire thing is mostly a minor aesthetic issue. We should still try to determine if it's a regression from for example https://github.com/koreader/koreader/pull/7415 or https://github.com/koreader/koreader/pull/7256 or if it's simply how it's been for five years without anyone noticing due to not speedracing to that specific menu after resume. ;-)

**Comment by matejdro** [comment]
I would argue it is also an usability issue, albeit a small one (I have to either wait for the timer to trigger or manually stop first timer before starting next one).

**Comment by Frenzie** [comment]
I don't recall there being a delay in the past.

**Comment by NiLuJe** [comment]
I'll look into it when I get back ;).

**Comment by NiLuJe** [comment]
* The UI scheduling is now on a monotonic timescale.
* CLOCK_MONOTONIC does *NOT* tick during suspend.
* This deals with the real (i.e., wall clock) timescale.

Guess what happens when you're trying to shove round pegs in a square hole? ;).

**Comment by NiLuJe** [comment]
This means that, say, I set a timer for one minute from now (through either option, doesn't matter), and I suspend, it'll effectively pop up ~1min *after* resume.

As @Frenzie said above, this is extremely useless.

I can just unschedule the timer on suspend and call it a day?

**Comment by Frenzie** [comment]
As a user I think that'd me feel insecure about its proper functioning. After all, 15 minutes have passed. Logically it should insta-fire on resume.

Possibly unschedule with notification?

**Comment by NiLuJe** [comment]
@Frenzie: A notification is tricky: we're going to sleep. It's easier going to show for a fraction of a second before the actual screensaver, or on top of it and possibly stay there depending on platforms ;).

**Comment by Frenzie** [comment]
Well, I'm not particularly interested in looking into it, but imo the old behavior was perfect. Do what you want. :-P

**Comment by NiLuJe** [comment]
Firing on resume is easily doable ;).
